In [1]:
import numpy as np

Question 1. 

In [2]:
A = [1, 2, 3, 4, 5, 6]
B = [13, 21, 34]

Aim is to form A_B = [1, 2, 3, 4, 5, 6, 13, 21, 34]

In [5]:
A_B = A.append(B)

In [7]:
print(A_B)

None


In [9]:
A_B = A.extend(B)

In [12]:
print(A_B)

None


We can see that both A.append(B) and A.extend(B) do not result in the desired answer

Question 2: Creating an identity matrix

In [13]:
np.array[(1,0,0), (0, 1, 0), (0, 0, 1)]

TypeError: 'builtin_function_or_method' object is not subscriptable

In [14]:
eye(3)

NameError: name 'eye' is not defined

In [15]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [16]:
np.array([1, 0, 0], [0, 1, 0], [0, 0, 1])

ValueError: only 2 non-keyword arguments accepted

We can see that only np.identity(3) generated an identity matrix

Question 3: Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [17]:
import pandas as pd

In [19]:
url = "https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv"

In [21]:
fuel_data = pd.read_csv(url)

In [22]:
fuel_data.columns

Index(['record_id', 'utility_id_ferc1', 'report_year', 'plant_name_ferc1',
       'fuel_type_code_pudl', 'fuel_unit', 'fuel_qty_burned',
       'fuel_mmbtu_per_unit', 'fuel_cost_per_unit_burned',
       'fuel_cost_per_unit_delivered', 'fuel_cost_per_mmbtu'],
      dtype='object')

In [30]:
fuel_data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].min()

fuel_type_code_pudl
coal         0.00
gas          0.00
nuclear      0.00
oil          0.00
other     -119.04
waste     -276.08
Name: fuel_cost_per_unit_burned, dtype: float64

We can see that waste has the lowest cost per unit burned of -276.08

Question 4: What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [31]:
fuel_data.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

We can see that the std of fuel_mmbtu_per_unit is 10.60 and the 75 percentile is 17.01

Question 5 : What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [32]:
fuel_data.kurtosis()

utility_id_ferc1                    1.088432
report_year                        -1.145656
fuel_qty_burned                   651.369450
fuel_mmbtu_per_unit                55.595695
fuel_cost_per_unit_burned         485.255851
fuel_cost_per_unit_delivered    11765.054226
fuel_cost_per_mmbtu             29489.132594
dtype: float64

In [37]:
fuel_data.skew()

utility_id_ferc1                  0.605070
report_year                       0.006953
fuel_qty_burned                  15.851495
fuel_mmbtu_per_unit               4.135217
fuel_cost_per_unit_burned        19.787440
fuel_cost_per_unit_delivered    105.014227
fuel_cost_per_mmbtu             171.675535
dtype: float64

We can see that the skewness and kurtosis of fuel quantity burned is 15.85 and 651.37 respectively

Question 6: Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places? 

In [38]:
fuel_data.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [39]:
len(fuel_data)

29523

In [48]:
percentage_missing = 180/len(fuel_data) *100
print('%.3f' %percentage_missing)

0.610


We can see that fuel_unit has 180 missing with 0.610%

Question 7: The feature with missing values falls under what category? What missing value imputation technique would you use?

In [50]:
fuel_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


In [51]:
fuel_data.fuel_unit.mode()

0    mcf
dtype: object

In [52]:
fuel_data.fuel_unit.mean()

TypeError: can only concatenate str (not "int") to str

We can see that mode is possible while mean is not possible. This means that since the data type is categorical, only categorical and mode imputations are possible

Question 8: Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [53]:
fuel_data.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

We can see that fuel_qty_burned and fuel_mmbtu_per_unit has the second and third lowest correlation with fuel_cost_per_unit_burned

Question 9: For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [85]:
fuel_data.groupby(['report_year', 'fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].mean()

report_year  fuel_type_code_pudl
1994         coal                     31.546467
             gas                      22.161183
             nuclear                3211.033155
             oil                      56.398759
             other                     9.728083
                                       ...     
2018         coal                     44.967758
             gas                       9.605542
             nuclear                8663.358050
             oil                     133.214969
             waste                    41.310750
Name: fuel_cost_per_unit_burned, Length: 148, dtype: float64

KeyError: '1994'

Question 10: Which year has the highest average fuel cost per unit delivered?

In [61]:
fuel_data.groupby('report_year')['fuel_cost_per_unit_delivered'].mean()

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

We can see that 1997 has the highest average fuel cost per unit delievered of 11140.20